### PR TITLE
Implement HuggingFace model server

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:
Support out-of-the-box serving for HuggingFace models

- all user need to do is to specify the huggingface model name or the storage uri which stores the model.
- By default execute preprocess, predict and postprocess steps in place, when predictor_host is specified the tensor inputs are sent to more optimized serving runtime like triton inference server following open inference protocol.
-  vllm can be used as the backend for the text generation case, will be implemented in a separate PR.

Screenshot 2024-01-03 at 9 04 25 AM

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [X] Sequence classification (bert-base-uncased-yelp-polarity)
- [X] Token classification (bert-large-cased-finetuned-conll03-english)
- [X] Text2Text generation (T5)
- [X] Fill mask (BERT)
- [X] Text Generation(BLOOM 7B)

**Special notes for your reviewer**:

1. This is the initial PR for the baseline HuggingFace inference runtime implementation, will need to test with more model types and add implementation for vLLM backend.

**Checklist**:

- [X] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

File: python/huggingfaceserver/huggingfaceserver/__main__.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import argparse
+import logging
+
+from huggingfaceserver import HuggingfaceModel, HuggingfaceModelRepository
+
+import kserve
+from kserve.errors import ModelMissingError
+
+DEFAULT_MODEL_NAME = "model"
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/__main__.py]: This is not used

File: python/kserve/kserve/utils/utils.py [hunk 2]
```diff
@@ -206,6 +208,8 @@ def get_predict_response(payload: Union[Dict, InferRequest], result: Union[np.nd
             infer_outputs=infer_outputs,
             response_id=payload.id if payload.id else generate_uuid()
         )
+    else:
+        raise ValueError("unsupported payload type")
```
Comment (terrytangyuan) [inline python/kserve/kserve/utils/utils.py]: Also print out the payload type to help debugging

File: python/kserve/kserve/protocol/infer_type.py [hunk 3]
```diff
@@ -112,6 +112,12 @@ def set_shape(self, shape):
         """
         self._shape = shape
 
+    def as_string(self) -> List[List[str]]:
+        if self.datatype == "BYTES":
+            return [s.decode("utf-8") for l in self._data for s in l]
+        else:
+            raise InvalidInput("invalid datatype in the input")
```
Comment (terrytangyuan) [inline python/kserve/kserve/protocol/infer_type.py]: Print out data type to help debugging

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+from kserve.protocol.infer_type import InferRequest, InferResponse, InferInput
+from kserve.utils.utils import get_predict_input, get_predict_response
+from kserve import Model
+import torch
+from transformers import AutoModelForCausalLM, AutoModelForSeq2SeqLM, AutoTokenizer, \
+    AutoConfig, \
+    AutoModelForSequenceClassification, AutoModelForTokenClassification, AutoModelForQuestionAnswering, \
+    AutoModelForMaskedLM, BatchEncoding
+
+ARCHITECTURES_2_TASK = {
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Consider moving to enum or constant class

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    "MTModel": "text2text-generation",
+    "EncoderDecoderModel": "text2text-generation",
+    # Model specific task for backward comp
+    "GPT2LMHeadModel": "text-generation",
+    "T5WithLMHeadModel": "text2text-generation",
+    "BloomModel": "text-generation",
+}
+
+
+def from_torch_dtype(torch_dtype):
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Consider defining a dictionary of mapping

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        model_config = AutoConfig.from_pretrained(model_config_path)
+        architecture = model_config.architectures[0]
+        task = None
+        for arch_options in ARCHITECTURES_2_TASK:
+            if architecture.endswith(arch_options):
+                task = ARCHITECTURES_2_TASK[arch_options]
+
+        if task is None:
+            raise ValueError(f"Task couldn't be inferred from {architecture}. Please manually set `task` option.")
+        return task
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py]: we can return task directly

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            return_tensors="pt",
+        )
+        context["input_ids"] = inputs["input_ids"]
+        # Serialize to tensor
+        if self.predictor_host:
+            infer_inputs = []
+            for key, input_tensor in inputs.items():
+                infer_input = InferInput(name=key, datatype=from_torch_dtype(input_tensor.dtype),
+                                         shape=list(input_tensor.shape), data=input_tensor.numpy())
+                print(infer_input.data)
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py]: I see that we are using both logging and print. Is that intentional?

**Review by terrytangyuan — COMMENTED** [review]
This is great! Left some comments after initial quick pass

File: python/huggingfaceserver/huggingfaceserver/model.py (hunk 4 above)
Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py]: created a task enum in task.py

**Comment by gavrissh** [comment]
This is great! lgtm for initial scenarios

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        else:
+            input_batch = self.tokenizer(
+                prompt,
+                max_length=self.max_length,
+                add_special_tokens=self.add_special_tokens,
+                return_tensors="pt",
+            )
+            if headers.get("streaming", "false") == "true":
+                streamer = AsyncGenerateStream(self.tokenizer)
+                generation_kwargs = dict(**input_batch, streamer=streamer)
```
Comment (cmaddalozzo) [inline python/huggingfaceserver/huggingfaceserver/model.py:178]:
```suggestion
                generation_kwargs = dict(**input_batch, **parameters, streamer=streamer)
```
Params provided in the request need to be passed to `generate`

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+                # TODO change to use thread pool executor
+                thread = Thread(target=self.model.generate, kwargs=generation_kwargs)
+                thread.start()
+                return streamer
+            else:
+                output_ids = self.model.generate(**input_batch)
```
Comment (cmaddalozzo) [inline python/huggingfaceserver/huggingfaceserver/model.py]:
```suggestion
                output_ids = self.model.generate(**input_batch, **parameters)
```

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+            self.kwargs["device_map"] = kwargs['device_map']
+        elif tp_degree > 0:
+            self.kwargs["device_map"] = "auto"
+            world_size = torch.cuda.device_count()
+            assert world_size == tp_degree, f"TP degree ({tp_degree}) doesn't match available GPUs ({world_size})"
+        self.model_id = kwargs.get('model_id', None)
+        self.model_dir = kwargs.get('model_dir', None)
+        self.do_lower_case = kwargs.get('do_lower_case', True)
+        self.add_special_tokens = kwargs.get('add_special_tokens', True)
+        self.max_length = kwargs.get('max_length', None)
```
Comment (gavrissh) [inline python/huggingfaceserver/huggingfaceserver/model.py:70]: Is it better to set this to a non `None` default value?

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: Hmm seems like there are both max_length and max_new_tokens , max length is going to be deprecated 
https://huggingface.co/docs/transformers/main_classes/text_generation

https://discuss.huggingface.co/t/confused-about-max-length-and-max-new-tokens/30892

Comment (gavrissh) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: yes, `max_new_tokens` is recommended now

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: Wait this arg is used for tokenizer and the max_length  default is none 

https://huggingface.co/transformers/v2.11.0/main_classes/tokenizer.html#:~:text=max_length%20(%20int%20%2C%20optional%20%2C%20defaults,the%20model%20with%20max_length%20%3D%20tokenizer.

Comment (gavrissh) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: Oh right, I confused it to the GenerationConfig param. This looks good!

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: It makes sense that they deprecated this for generate config to avoid confusion.

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:70]: So we can add max_new_token when you work on vllm

**Review by cmaddalozzo — CHANGES_REQUESTED** [review]

**Comment by FlyAIBox** [comment]
Great, when can it be applied to the production environment, expect @yuzisun @terrytangyuan

**Comment by yuzisun** [comment]
> Great, when can it be applied to the production environment, expect @yuzisun @terrytangyuan

I think we will need to wait for another release, currently it’s alpha state for people to try out and give feedback.

File: python/kserve/kserve/protocol/rest/v2_datamodels.py [hunk 11]
```diff
@@ -248,3 +248,83 @@ class Config:
...
+    class Config:
+        json_loads = orjson.loads
+        schema_extra = {
+            "example": {
+                "text_output": "Tell me about the AI",
+                "model_name": "bloom7b1",
+                "outputs": [
+                    {
+                        "id": "267",
+                        "logprod": -2.0723474,
```
Comment (gavrissh) [inline python/kserve/kserve/protocol/rest/v2_datamodels.py]: typo - logprob?

File: python/huggingfaceserver/README.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      protocolVersion: v2
+      resources:
+        limits:
+          cpu: "1"
+          memory: 8Gi
+        requests:
+          cpu: "1"
+          memory: 8Gi
+          nvidia.com/gpu: "1"
+      runtimeVersion: 23.10-py3
```
Comment (zoramt) [inline python/huggingfaceserver/README.md:86]: This is newer than what we have in v0.11.2. Curious why you're choosing 23.10 vs 23.05.

Comment (yuzisun) [reply python/huggingfaceserver/README.md:86]: no specific reason, either version should work fine. KServe sets a default version which we use to run e2e tests, but you can always get onto newer version by specifying `runtimeVersion`

Comment (zoramt) [reply python/huggingfaceserver/README.md:86]: Thanks!

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        else:
+            input_batch = self.tokenizer(
+                prompt,
+                max_length=self.max_length,
+                add_special_tokens=self.add_special_tokens,
+                return_tensors="pt",
+            )
+            if headers.get("streaming", "false") == "true":
+                streamer = AsyncGenerateStream(self.tokenizer)
+                generation_kwargs = dict(**input_batch, **parameters, treamer=streamer)
```
Comment (cmaddalozzo) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Typo
```suggestion
                generation_kwargs = dict(**input_batch, **parameters, streamer=streamer)
```

File: .github/workflows/python-test.yml [hunk 14]
```diff
@@ -221,3 +221,29 @@ jobs:
           cd python
           source alibiexplainer/.venv/bin/activate
           pytest --cov=alibiexplainer ./alibiexplainer
+
+      # ----------------------------------------Huggingface Server Unit Tests------------------------------------------------
+      # @TODO this needs github action large runner
```
Comment (terrytangyuan) [inline .github/workflows/python-test.yml]: This is new and might help here https://github.blog/2024-01-17-github-hosted-runners-double-the-power-for-open-source/

Comment (yuzisun) [reply .github/workflows/python-test.yml]: awesome! going to try out!

Comment (yuzisun) [reply .github/workflows/python-test.yml]: nice looks like now I can run the hugging face tests

Comment (terrytangyuan) [reply .github/workflows/python-test.yml]: Great!

**Comment by johnugeorge** [comment]
Thanks @yuzisun 

/lgtm
